### ISSUE TITLE
SideNav and HeaderContainer are both trying to control the same thing

### ISSUE THREAD

**Issue by aledavila** [carbon-design-system/carbon#3223]
`Side Nav` and `Header Container` in UI-Shell react need to be refactored. As they stand they are overriding each other over control of the `side-nav-expanded`. Tried using ReactContext but could not do it as the correct values aren't being passed in. There are 2 different functions controlling the expand.

**Comment by vpicone** [comment]
Me and @abbeyhrt are gonna give this a go using Context again.

**Comment by asudoh** [comment]
@aledavila My knee-jerk reaction to this issue is that we can simply let header container control side nav - Meaning the source of truth is header container if both exist in an app. Or did you have any reason you think that's not feasible?

**Comment by vpicone** [comment]
@asudoh the header container will still be the source of truth, passing the expanded state into the sidenav menus and sidenav items however is the challenge since they are supplied by the dev and we just pass them through as children.

We could loop over the children in sidenav and attach the prop, looping and applying prop updates in this way gets increasingly complex as we add pieces of state such as the right panel, the search, and the perhaps things like focus state. React context gives us the opportunity to pass the state into children and nested children without looping over the children imperatively. We'd not only need to loop over the sidenav children but sidenav menu children as well to apply the expanded state to children of the menu items.

Another option is requiring something like getSideNavMenuProps, however this would be a breaking change and significantly complicate the usage of the shell components. Context allows the state to be shared to any children of the provider which would live in HeaderContainer.

**Comment by aledavila** [comment]
This is the issue for make the side nav sub menu collapse: https://github.com/carbon-design-system/carbon/issues/3222

Just linking it for reference.

**Comment by asudoh** [comment]
@vpicone Would you be able to put a code snippet of how below code would be _without_ the context adding the feature of #3222, assuming that the notion of the context is for #3222? If it's not the case, could you elaborate what feature the notion of the context is for? Thanks a lot!

```javascript
<HeaderContainer
  render={({ isSideNavExpanded, onClickSideNavExpand }) => (
    <>
      ...
      <SideNav
        aria-label="Side navigation"
        expanded={isSideNavExpanded}>
        <SideNavItems>
          <SideNavMenu renderIcon={Fade16} title="Category title">
            <SideNavMenuItem href="javascript:void(0)">
              Link
            </SideNavMenuItem>
            <SideNavMenuItem href="javascript:void(0)">
              Link
            </SideNavMenuItem>
            ...
          </SideNavMenu>
          <SideNavMenu renderIcon={Fade16} title="Category title">
            <SideNavMenuItem href="javascript:void(0)">
              Link
            </SideNavMenuItem>
            <SideNavMenuItem href="javascript:void(0)">
              Link
            </SideNavMenuItem>
            ...
          </SideNavMenu>
          ...
        </SideNavItems>
      </SideNav>
      ...
    </>
  )}
/>
```

**Comment by vpicone** [comment]
@asudoh 

In order to get this to work without context, we would need to force users to included the expanded state as well as function to update the state in the parent container. This is a breaking change and it introduces a lot of complexity to the public component API (as you can see below). It would also be a breaking change, the menus would not function properly without these additional props.

The advantage to context is that there would be no changes for the developer. The code you pasted above would just work as intended. The state and update functions are handled behind the scenes through context as opposed to being imperatively applied to menu children.

---

**By far the biggest breaking change*: would be that SideNav would ALSO need to be changed to a renderProp pattern. Since this component can be controlled, the SideNavMenu's would need to be aware of the state coming from this source as well. 

To simplify, we could make SideNav completely uncontrolled and insist it lives in a HeaderContainer (which seems ridiculous If they don't want a header), then make it completely uncontrolled. But these are both breaking changes and make for a really awkward API.

Here's an example of how it would need to be without context. SideNav would need a way to toggle the HeaderContainer state and also be controlled by it. the SideNavMenu items would need to know whether or not the SideNav was expanded. This would come from a resolution in SideNav that determines which piece of the shell is controlling the expand state and then passed down into the menus via render props.

Here's an example of what the usage of UIShell would look like if we wanted to allow SideNav to both be controllable and used in isolation while also allowing it to be controlled by the parent HeaderContainer

```
<HeaderContainer
  render={({ isSideNavExpanded, onClickSideNavExpand }) => (
    <>
      ...
      <SideNav
        aria-label="Side navigation"
        expanded={isSideNavExpanded}
		onClickSideNavExpand={onClickSideNavExpand}
        render={({ isSideNavExpandedForReal }) => (
        <SideNavItems>
          <SideNavMenu isSideNavExpanded={isSideNavExpandedForReal} renderIcon={Fade16} title="Category title">
            <SideNavMenuItem href="javascript:void(0)">
              Link
            </SideNavMenuItem>
            <SideNavMenuItem href="javascript:void(0)">
              Link
            </SideNavMenuItem>
            ...
          </SideNavMenu>
          <SideNavMenu isSideNavExpanded={isSideNavExpandedForReal} renderIcon={Fade16} title="Category title">
            <SideNavMenuItem href="javascript:void(0)">
              Link
            </SideNavMenuItem>
            <SideNavMenuItem href="javascript:void(0)">
              Link
            </SideNavMenuItem>
            ...
          </SideNavMenu>
          ...
        </SideNavItems>

        )
        >
      </SideNav>
      ...
    </>
  )}
/>
```

**Comment by asudoh** [comment]
Thanks @vpicone for elaborating! My question give your reply above is why we need `isSideNavExpandedForReal`? I thought we can just do below, given we are controlling `<SideNav>` with `isSideNavExpanded` prop and thus `<SideNav>`'s actual expanded state always matches `isSideNavExpanded` prop:

```javascript
<SideNavMenu isSideNavExpanded={isSideNavExpanded} renderIcon={Fade16} title="Category title">
```

**Comment by vpicone** [comment]
Ah I see! I have a couple questions with this approach. 

1. How do you share nav state with menus?
2. How would you share the nav state to menus if there is no Header? If SideNav is in isolation. 
3. How do you toggle the SideNav state when the user hovers or focuses the side nav rail?

These are the problems which lead me towards context. It allows us to accomplish them without breaking changes/prop getters.

**Comment by asudoh** [comment]
@vpicone Cool - If I understand your question:

1. The menus can also refer to `isSideNavExpanded`.
2. We can still use `<HeaderContainer>` if side nav state has to be shared with other components. We can rename `<HeaderContainer>` to `<NavContainer>` or pick up any name, given `<HeaderContainer>` is not associated with any user interface.
3. `<SideNav>` can send an event via `onClickSideNavExpand` or a new event. `<HeaderContainer>` can listen to that and update `isSideNavExpanded` or a new state (render prop). Below is the approach using an existing event/state/render-prop:

```javascript
<HeaderContainer
  render={({ isSideNavExpanded, onClickSideNavExpand }) => (
    <>
      ...
      <SideNav
        aria-label="Side navigation"
        expanded={isSideNavExpanded}
        onClickSideNavExpand={onClickSideNavExpand}>
        ...
      </SideNav>
      ...
    </>
  )}
/>
```

Below is the approach using a new pair of event/state/render-prop (though we have a separate question, which is, whether the hover state of `<SideNav>` should be synced to other components):

```javascript
<HeaderContainer
  render={({ isSideNavExpanded, isSideNavHovered, onClickSideNavExpand, onHoverSideNav }) => (
    <>
      ...
      <SideNav
        aria-label="Side navigation"
        expanded={isSideNavExpanded}
        hovered={isSideNavHovered}
        onClickSideNavExpand={onClickSideNavExpand}
        onHOverSideNav={onHoverSideNav}>
        ...
      </SideNav>
      ...
    </>
  )}
/>
```

**Comment by vpicone** [comment]
Right, so all of those things require devs to apply the correct props to all of the components that need it (prop getters probably?). If you apply your solution to even one of the smaller examples in the stories you’ll see how crazy the props get. Your change also means the SideNav would need to be uncontrolled and nested, otherwise it can’t share the expansion events with its children menus. Like I’ve said, these would all also be breaking changes. 

The advantage of context is that we don’t rely on the developers using prop getters or manually applying the correct props to provide the essential information our components need to render properly. In addition, components can exist in isolation or share a context provider, all without complicating the normal API with props that we have no intention of developers controlling/updating. 

 You can also imagine how things could get complicated with a right panel like we have in the website. Expansion of that element collapses the side nav in the website.

**Comment by vpicone** [comment]
I feel like this is the exact use case the React team had in mind for context. I hope I’ve done a good job of explaining the advantages to using context here, do you mind sharing what you perceive as the benefits for the render prop approach?

**Comment by asudoh** [comment]
The intent of the render prop approach is not needing to change our shell components (e.g. `<SideNav>`) and not to add any dependency (to shared variables) in our shell components, for the sake of less need of prop drilling/propagation. Such added dependency will reduce reusability/portability of `<SideNav>`, etc., as PR# 1085 in [reactjs.org repo](https://github.com/reactjs/reactjs.org) alludes to. As of today, `<HeaderContainer>` is _not_ needed to use `<SideNav>`, which means `<SideNav>` can be used anywhere without being concerned of its dependency. Therefore, another approach would be just letting application to use context as opposed to baking in the context approach to any of our shell component.

**Comment by asudoh** [comment]
@vpicone BTW could you elaborate on why `<SideNavMenu>` (in addition to `<SideNav>`) needs to have `isSideNavExpanded` prop? Thanks!

**Comment by vpicone** [comment]
@asudoh Render props can mean less prop drilling, but for us here it means more props that the devs must apply for components to render properly. You also have to pass the props given to the render function down to the children somehow. It's certainly not an advantage here compared to context which doesn't require devs to supply the correct props just so our components can function. It also complicates the API because we're asking them to supply the same prop to every menu but not control it themselves?

We'd also need to change the components to toggle the state of their parent component. This is yet another prop we make devs supply otherwise the component won't function properly.

Context allows you to supply that state and means of updating it without imperatively passing the correct props down to each child (which for us would be passing the burden onto the developer). I believe the portability Dan talks about in [that article](https://reactjs.org/docs/context.html#before-you-use-context) is referring to use cases involving business logic such as theme or user data. Since that context is often critical to the functioning of the component, it can't be placed in a new library or app without a provider that can supply that context.

We get around that here by 
1. Shipping the context provider with the library (HeaderContainer or w/e it's renamed to. 
2. using default context values that would allow each individual component to be used in isolation without the context provider 


re: sidenav menus receiving the expanded state

As of today, SideNav is "portable", but that's only because it doesn't need to share it's state with SideNav menu children. #3222 requires that the menus need to have isSideNavExpanded in order to collapse. 

Here are the alternatives as I see them and my concerns:

We could force developers to use the wrapper and manage state in the wrapper (component doesn't function fully in isolation, prescriptive markup, breaking change)

We could imitate a collapsed state with css but then we are controlling app behavior with styles (multiple sources of truth: react thinks menu is expanded, styles suggest they are not).

**Comment by asudoh** [comment]
To make the story short, once a component depend on a React context, a similar thing to global variable dependency happens. That said, any approach that avoids such problem will be good.

**Comment by vpicone** [comment]
@asudoh the default context value allows a fallback if the component is not within a context provider. Could you rephrase your last sentence?

**Comment by asudoh** [comment]
> Could you rephrase your last sentence?

Are you referring to:

> That said, any approach that avoids such problem will be good.

Or something else?

**Comment by vpicone** [comment]
@asudoh yes that bit! Would you be okay with me going forward with a context approach as apposed to render props?

**Comment by asudoh** [comment]
@vpicone Not baking in React context dependency to the UI components for UI shell (`<SideNav>`, etc) is the only imperative. That said, if we are just creating a new non-UI-rendering React component that works with React context it's OK.

**Comment by vpicone** [comment]
@asudoh the default value for context means that the components will have a fallback if they aren't wrapped in a context provider. If there is no context provider, they use the default context value. So in this way there is no external dependency from the perspective of the components.

**Comment by asudoh** [comment]
That's good, and to extend that, we need to separate the UI components (e.g. `<SideNav>`) with the context-aware component, to align to our approach of creating components that does one thing and do it well and let them work together.

**Comment by joshblack** [comment]
Going through issues, added this to our release planning for next release @aledavila 👀  Closing for now until we know what this update will look like. Let me know if I should re-open!
